app/models/qernel/plugins/merit/import_adapter.rb

```diff
@@ -1,64 +1,36 @@
 module Qernel::Plugins
   module Merit
-    class ImportAdapter < Adapter
+    # Implements behaviour specific to the import interconnector.
+    class ImportAdapter < ProducerAdapter
       def inject!
+        super
+
         elec_link = @converter.converter.output(:electricity).links.first
-        demand    = participant.production(:mj)
 
         if elec_link.link_type == :flexible
           # We need to override the calculation of the flexible link and set
           # set the demand explicitly.
-          elec_link.dataset_set(:value, demand)
+          elec_link.dataset_set(:value, @converter.demand)
           elec_link.dataset_set(:calculated, true)
         end
-
-        @converter.demand = demand
       end
 
       private
 
-      def producer_attributes
-        attrs = super
-
-        # The marginal cost of the import producer is 1 EUR more than the the
-        # most expensive participant.
-        #
-        # See https://github.com/quintel/merit/issues/143
-        attrs[:marginal_costs]           = max_marginal_cost + 1.0
-        attrs[:number_of_units]          = 1.0
-        attrs[:output_capacity_per_unit] = @converter.electricity_output_capacity
-
-        attrs
-      end
-
       def producer_class
         ::Merit::DispatchableProducer
       end
 
-      # Internal: Determines the marginal cost of the most expensive merit order
-      # participant.
-      #
-      # Returns a Float.
-      def max_marginal_cost
-        types = @graph.plugin(:merit).participant_types
-
-        mo_converters = @graph.converters.select do |conv|
-          conf = conv.dataset_get(:merit_order)
+      def marginal_costs
+        @graph.carrier(:imported_electricity).cost_per_mj * 3600
+      end
 
-          conf &&
-            conf.type != :consumer &&
-            conv.query.number_of_units > 0 &&
-            types.include?(conf.type)
-        end
+      def output_capacity_per_unit
+        @converter.electricity_output_capacity
+      end
 
-        mo_converters.map do |conv|
-          begin
-            cost = conv.query.marginal_costs
-            cost.nan? || cost == Float::INFINITY ? 0.0 : cost
-          rescue
-            0.0
-          end
-        end.max
+      def flh_capacity
```

Comment:
Here flh_capacity is defined as the electricity_output_capacity of the converter. This means that the demand of the node is based on its output. This is only correct if losses are zero. I don't understand this. Perhaps the name flh_capacity can be chosen to be more specific?

Reply:
This is only used by the import interconnector. Sliders adjust the output capacity of this converter; it doesn't have an input capacity.